### PR TITLE
Auth.provider

### PR DESCRIPTION
**Pull request by mozzie2004**

### CODE REVIEW

File: packages/frontend/helpers/auth.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    return this._storage.get(StorageKey.ACCESS_TOKEN);
+  }
+
+  public getRefreshToken(): string {
+    return this._storage.get(StorageKey.REFRESH_TOKEN);
+  }
+
+  public setTokens(accessToken: string, refreshToken: string): void {
+    this._storage.set(StorageKey.ACCESS_TOKEN, accessToken);
+    this._storage.set(StorageKey.REFRESH_TOKEN, refreshToken);
```
Comment (Nightlngale) [inline packages/frontend/helpers/auth.ts:21]: add httpOnly option

File: packages/frontend/helpers/auth.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import type { Auth, Storage } from '@vse-bude/shared';
+import { StorageKey } from '@vse-bude/shared';
+
+class AuthHelper implements Auth {
```
Comment (Nightlngale) [inline packages/frontend/helpers/auth.ts:4]: We definitely need a better name for Auth

File: packages/frontend/pages/index.tsx [hunk 3]
```diff
@@ -3,16 +3,15 @@ import { useTranslation } from 'next-i18next';
...
 import { useAppDispatch, useTypedSelector } from '@hooks';
 import { shallowEqual } from 'react-redux';
 import { fetchRandomData, fetchRandomDataSSR, wrapper } from 'store';
 
 export const getServerSideProps = wrapper.getServerSideProps(
   (store) => async (ctx) => {
-    const storage = new CookieStorage(ctx);
     const { locale } = ctx;
-    const httpClient = new Http(process.env.NEXT_PUBLIC_API_ROUTE, storage);
+    const httpClient = new Http(process.env.NEXT_PUBLIC_API_ROUTE, auth);
```
Comment (Nightlngale) [inline packages/frontend/pages/index.tsx:14]: will we have a access token in auth?

File: packages/frontend/providers/AuthProvider/index.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+  useEffect(() => {
+    if (needLoadUser) {
+      dispatch(getCurrentUser());
+    }
+  }, [dispatch, needLoadUser, router]);
+
+  useEffect(() => {
+    if (!hasToken) {
+      router.push('/');
```
Comment (Nightlngale) [inline packages/frontend/providers/AuthProvider/index.tsx:32]: use Enum Routes

File: packages/frontend/providers/AuthProvider/index.tsx [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+const AuthProvider = ({ children }: AuthProviderProps) => {
+  const user = useTypedSelector((state) => state.profile.user);
+
+  const dispatch = useAppDispatch();
+  const router = useRouter();
+
+  const hasToken = !!authHelper.getAccessToken();
+  const hasUser = !!user;
+  const needLoadUser = !hasUser && hasToken;
```
Comment (Nightlngale) [inline packages/frontend/providers/AuthProvider/index.tsx:22]: needToLoadUser

File: packages/frontend/store/profile/actions.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import { getUser } from '@services';
+import { ProfileActions } from './action-types';
+
+const logout = createAction(ProfileActions.LOG_OUT);
+
+export const getCurrentUser = createAsyncThunk(
+  ProfileActions.LOG_IN,
+  async (_request, { dispatch, rejectWithValue }) => {
+    try {
+      return await getUser();
```
Comment (Nightlngale) [inline packages/frontend/store/profile/actions.ts:12]: no need to await

File: packages/frontend/store/profile/actions.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    } catch (err) {
+      const isHttpError = err instanceof HttpError;
+
+      if (isHttpError && err.status === HttpStatusCode.UNAUTHORIZED) {
+        dispatch(logout());
+      }
```
Comment (Nightlngale) [inline packages/frontend/store/profile/actions.ts:18]: This case should be handled in httpService

File: packages/frontend/store/profile/reducer.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    },
+    [getCurrentUser.rejected.type](state) {
+      state.user = null;
+      state.loading = false;
+    },
+    [HYDRATE](state, { payload }: HydrateAction) {
+      if (payload.profile.user) {
+        state.user = payload.profile.user;
+      }
+    },
```
Comment (Nightlngale) [inline packages/frontend/store/profile/reducer.ts:37]: Not sure if we need to load user in SSR

File: packages/shared/src/helpers/https.ts [hunk 9]
```diff
@@ -119,24 +121,48 @@ class Http {
...
 
     return result.json() as Promise<T>;
   }
 
-  private updateAuthorizationToken() {
-    throw new Error('not implemented');
+  private async updateAuthorizationToken() {
+    const refreshToken = this._auth.getRefreshToken();
+
+    if (refreshToken) {
```
Comment (Nightlngale) [inline packages/shared/src/helpers/https.ts:145]: what if no refresh token?

**Review by Nightlngale — CHANGES_REQUESTED** [review]

File: packages/frontend/helpers/auth.ts (hunk 1 above)
Comment (Nightlngale) [inline packages/frontend/helpers/auth.ts:21]: Need httpOnly option

File: packages/frontend/pages/index.tsx (hunk 3 above)
Comment (mozzie2004) [reply packages/frontend/pages/index.tsx:14]: no, i didn't think about it. I will do it

File: packages/shared/src/common/enums/api/routes.ts [hunk 10]
```diff
@@ -9,4 +9,5 @@ export enum AuthApiRoutes {
   SIGN_UP = '/sign-up',
   SIGN_OUT = '/sign-out',
   REFRESH_TOKEN = '/refresh-token',
+  GET_USER = '/user',
```
Comment (Nightlngale) [inline packages/shared/src/common/enums/api/routes.ts:12]: USER = '/user'

**Review by Nightlngale — APPROVED** [review]

**Pull request merged**